### DISCUSSION TITLE
Custom problem normalizer ignored for GET requests

### DISCUSSION THREAD

**Discussion by betzda1** [symfony/symfony#57993] [Q&A] [answered]
I like **all** my errors to be in JSON and have a `message` property with error details. Thats why I followed [this documentation](https://symfony.com/doc/current/controller/error_pages.html#overriding-error-output-for-non-html-formats) and implemented a custom problem normalizer (SF 7.1).

It's basically outputting an exception's message as `message` property in a JSON response.

```php
<?php

namespace App\Serializer;

use InvalidArgumentException;
use Symfony\Component\ErrorHandler\Exception\FlattenException;
use Symfony\Component\Serializer\Normalizer\NormalizerInterface;

class CustomProblemNormalizer implements NormalizerInterface
{
    public function getSupportedTypes(?string $format): array
    {
        return [
            FlattenException::class => __CLASS__ === self::class,
        ];
    }

    public function normalize(mixed $object, ?string $format = null, array $context = []): array
    {
        if (!$object instanceof FlattenException) {
            throw new InvalidArgumentException(sprintf('The object must implement "%s".', FlattenException::class));
        }

        return [
            'message' => $object->getMessage()
        ];
    }

    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
    {
        return $data instanceof FlattenException;
    }
}
```
If I test this against a simple controller as such:

```php
<?php

declare(strict_types=1);

namespace App\Controller;

use App\Entity\UserProfile;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Attribute\Route;

class UserProfileController extends AbstractController
{
    #[Route('/api/user-profiles/{uuid}', methods: ['GET'])]
    public function getUserProfileByUuid(Request $request, UserProfile $userProfile): JsonResponse
    {
        return new JsonResponse(["success" => true]);
    }

    #[Route('/api/user-profiles/{uuid}', methods: ['PUT'])]
    public function updateUserProfile(Request $request, UserProfile $userProfile): JsonResponse
    {
        return new JsonResponse(["success" => true]);
    }
}
```

I receive the following results. `PUT` is fine:
```json
PUT /api/user-profiles/foobar
{
	"message": "\"App\\Entity\\UserProfile\" object not found by \"Symfony\\Bridge\\Doctrine\\ArgumentResolver\\EntityValueResolver\"."
}
```
`GET` ignores my normalizer:
```json
GET /api/user-profiles/foobar
```
![image](https://github.com/user-attachments/assets/039f22c2-cf30-49a4-b022-5afd24a22dc9)

It does not matter if I use the `EntityValueResolver` or if I throw the `NotFoundHttpException` myself. It seems to be dependent of the HTTP verb.

Does anyone know

1. If this is by design or maybe worth an issue?
2. If it is by design: how to work around this behavior for my `GET` endpoints?

**Comment by MatTheCat** [comment, marked as answer]
When the documentation titles `Overriding Error output for non-HTML formats`, the format is the request’s.

You can set it either by passing the ~`Content-Type`~ `Accept` header in your request, or by configuring [the special `format` routing parameter](https://symfony.com/doc/current/routing.html#special-parameters).

**Reply by betzda1** [reply]
Indeed, I got fooled by a (automatically set) `Content-Type` header in the PUT request. Setting `Content-Type` or `Accept` also for the `GET` also activates the `CustomProblemNormalizer` for the `NotFoundHttpException`. Thanks for pointing me into the right direction!
